### PR TITLE
Allow MicroShift to join new worker nodes

### PR DESCRIPTION
**Pull request by oglok**
Allow MicroShift to join new worker nodes, according to design here #498  (see individual commits for review),

* 4523987f Add flags to allow TLS bootstrapping of nodes
* 9d8e2ade Add bootstrap module and generate token file
* 26f04a33 Add ClusterRoleBinding for bootstrap process
* 1103d731 Generate bootstrap kubeconfig
* 1fd2a3c9 Allow MicroShift to start node role standalone
* 6c84f168 Apply CRB for bootstraping nodes
* 1366626f Use bootstrap kubeconfig for kube-proxy
* c2eacfd6 Use netcgo insted of netgo
* 40bff393 Add vagrant env to test/devel/debug multi-worker

Related PRs: #499 , #500

### CODE REVIEW

**Comment by oglok** [comment]
For some reason, the cluster role bindings added using bindata strategy are not being applied. That causes kubelet to:

```
E1125 16:20:51.785446   27509 server.go:294] "Failed to run kubelet" err="failed to run Kubelet: cannot create certificate signing request: certificatesigningrequests.certificates.k8s.io is forbidden: User \
"kubelet-bootstrap\" cannot create resource \"certificatesigningrequests\" in API group \"certificates.k8s.io\" at the cluster scope" 
```

**Comment by oglok** [comment]
Now the ClusterRoleBindings get properly applied.

**Comment by oglok** [comment]
https://asciinema.org/a/452317

File: pkg/node/kube-proxy.go [hunk 1]
```diff
@@ -87,14 +87,12 @@ func (s *ProxyOptions) writeConfig(cfg *config.MicroshiftConfig) error {
...
-	}
-data = append(data, `
-hostnameOverride: ` + cfg.NodeName + `
-clusterCIDR: ` + cfg.Cluster.ClusterCIDR + `
+
+	data = append(data, "\n"+`  kubeconfig: `+cfg.DataDir+"/resources/kubelet/kubeconfig"...)
+
+	data = append(data, `
+hostnameOverride: `+cfg.NodeName+`
+clusterCIDR: `+cfg.Cluster.ClusterCIDR+`
```
Comment (mangelajo) [inline pkg/node/kube-proxy.go:90]: Move this to the previous commit.

File: pkg/cmd/init.go [hunk 2]
```diff
@@ -17,14 +17,24 @@ package cmd
...
 )
 
+func mustInitAll(cfg *config.MicroshiftConfig) {
+	err := initAll(cfg)
+	if err != nil {
+		klog.ErrorS(err, "error during initAll")
+		os.Exit(1)
+	}
+}
+
```
Comment (mangelajo) [inline pkg/cmd/init.go:37]: we can use: util.Must

File: pkg/cmd/run.go [hunk 3]
```diff
@@ -63,9 +64,18 @@ func RunMicroshift(cfg *config.MicroshiftConfig, flags *pflag.FlagSet) error {
...
 	// TODO: change to only initialize what is strictly necessary for the selected role(s)
-	if _, err := os.Stat(filepath.Join(cfg.DataDir, "certs")); errors.Is(err, os.ErrNotExist) {
-		if config.StringInList("controlplane", cfg.Roles) {
-			initAll(cfg)
+	if config.StringInList("controlplane", cfg.Roles) {
+		if _, err := os.Stat(filepath.Join(cfg.DataDir, "certs")); errors.Is(err, os.ErrNotExist) {
+			mustInitAll(cfg)
+		} else {
+			err = loadCA(cfg)
+			if err != nil {
```
Comment (mangelajo) [inline pkg/cmd/run.go:72]: ???? is this the right thing to do?

File: pkg/controllers/kube-apiserver.go [hunk 4]
```diff
@@ -117,14 +117,11 @@ func (s *KubeAPIServer) configure(cfg *config.MicroshiftConfig) {
 		"--tls-private-key-file=" + cfg.DataDir + "/certs/kube-apiserver/secrets/service-network-serving-certkey/tls.key",
 		"--token-auth-file=" + cfg.DataDir + "/resources/microshift-bootstrap-token",
```
Comment (mangelajo) [inline pkg/controllers/kube-apiserver.go:118]: Move this to a separate PR!

File: Makefile [hunk 5]
```diff
@@ -71,7 +71,8 @@ debug:
 	@echo SOURCE_GIT_TAG:"$(SOURCE_GIT_TAG)"
 
 # These tags make sure we can statically link and avoid shared dependencies
-GO_BUILD_FLAGS :=-tags 'include_gcs include_oss containers_image_openpgp gssapi providerless netgo osusergo'
+# netcgo : use system resolver for DNS instead of the netgo implementation which does not include mDNS
+GO_BUILD_FLAGS :=-tags 'include_gcs include_oss containers_image_openpgp gssapi providerless netcgo osusergo'
```
Comment (mangelajo) [inline Makefile:75]: explain in commit message.

File: pkg/util/cert.go [hunk 6]
```diff
@@ -72,6 +72,43 @@ func GenCA(common string, svcName []string, duration time.Duration) (*rsa.Privat
...
+	}
+
+	certPath := filepath.Join(dir, certFilename)
+	cert, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return errors.Wrap(err, "reading CA certificate")
+	}
+
+	block, _ = pem.Decode(cert)
+	rootCA, err = x509.ParseCertificate(block.Bytes)
```
Comment (oglok) [inline pkg/util/cert.go:98]: util.cert has a function called PemToCertificate that might be good to use.

File: pkg/util/cert.go [hunk 7]
```diff
@@ -72,6 +72,43 @@ func GenCA(common string, svcName []string, duration time.Duration) (*rsa.Privat
...
+	if err != nil {
+		return errors.Wrap(err, "error reading CA key")
+	}
+
+	block, _ := pem.Decode(key)
+	if block == nil {
+		return errors.New("failed to parse PEM block containing the CA key")
+	}
+
+	rootKey, err = x509.ParsePKCS1PrivateKey(block.Bytes)
```
Comment (oglok) [inline pkg/util/cert.go:86]: util.cert has a function called PemToPrivateKey that might be good to use.

**Review by oglok — COMMENTED** [review]
Good job! Some minor comments. Once the klog patch is out of the PR, we can remove the WIP so it can be reviewed.

File: pkg/util/cert.go (hunk 7 above)
Comment (mangelajo) [reply pkg/util/cert.go:86]: oh nice, makes sense

File: pkg/util/cert.go (hunk 6 above)
Comment (mangelajo) [reply pkg/util/cert.go:98]: makes sense

Comment (mangelajo) [reply pkg/util/cert.go]: Done.

File: pkg/util/cert.go (hunk 7 above)
Comment (mangelajo) [reply pkg/util/cert.go]: Done.

**Comment by fzdarsky** [comment]
@oglok @mangelajo this is a pretty big PR and conflates many changes (introducing debug flags, formatting changes, config param changes, mDNS, multi-node, etc.) into one, which makes it hard to review. Can we split this into independent PRs?

**Comment by fzdarsky** [comment]
Also, as this impacts the MicroShift architecture and UX, we'll need an RFE doc in `/docs/design` that describes the design trade-offs and proposed solution before that solution can be implemented.

**Comment by mangelajo** [comment]
@fzdarsky its broken into logical commits, we expected that would suffice for agility.

If we go the PR route this is going to be eternal.

I'm ok with extracting specific commits to individual PRs but I don't think is a good approach to split all since those are dependent (if we want this for 2021) I'm fine if we are ok with slipping off Quarter

**Comment by mangelajo** [comment]
In gerrit chained patches are easier but GitHub is a mess when you try to chain dependent PRs

**Comment by mangelajo** [comment]
I will split out some patches which can be out of the chain and topic of this PR without great issue, the only thing is that for this PR to be functional (or if somebody wants to test the PR) those other patches will need to be merged first

**Comment by mangelajo** [comment]
Extracted to separate PR: https://github.com/redhat-et/microshift/pull/497

**Comment by mangelajo** [comment]
@oglok @fzdarsky  I hope it looks better now, I'm going to review the RFE

**Comment by oglok** [comment]
Closing for now until we decide the multi-node strategy. This is a good PoC for learning to build the right solution.

**Pull request closed without merging**